File: code/day12.jl
function main()
    code = readlines("data/day12")
    ship(code), waypoint(code)
end

function ship(code)
    coord = Dict("E"=>0, "N"=>0, "S"=>0, "W"=>0)
    facing = Dict(0=>"E", 90=>"N", 180=>"W", 270=>"S")
    angle, line = 0, 1
    for line in eachindex(code)
        command, moves = match(r"([A-Z])(.+)", code[line]).captures
        moves = parse(Int, moves)
        if command == "L"
            angle = mod((angle + moves), 360)
        elseif command == "R"
            angle = mod((angle - moves), 360)
        elseif command == "F"
            coord[facing[angle]] += moves
        else
            coord[command] += moves
        end
    end
    abs(coord["N"] - coord["S"]) + abs(coord["E"] - coord["W"])
end

function waypoint(code)
    wpx, wpy = 10, 1
    spx, spy = 0, 0
    for line in eachindex(code)
        command, moves = match(r"([A-Z])(.+)", code[line]).captures
        moves = parse(Int, moves)
        if command == "N"
            wpy += moves
        elseif command == "S"
            wpy -= moves
        elseif command == "E"
            wpx += moves
        elseif command == "W"
            wpx -= moves
        elseif command ∈ ["R", "L"] && moves == 180
            wpx, wpy = -wpx, -wpy
        elseif command == "R"
            if moves == 90
                wpx, wpy = wpy, -wpx
            elseif moves == 270
                wpx, wpy = -wpy, wpx
            end
        elseif command == "L"
            if moves == 90
                wpx, wpy = -wpy, wpx
            elseif moves == 270
                wpx, wpy = wpy, -wpx
            end
        else
            spx += moves * wpx
            spy += moves * wpy
        end
    end
    abs(spx) + abs(spy)
end

File: code/day12.R
readLines("data/day12")
gsub("[A-Z]", "", "F10")
gsub("[0-9]+", "", "F10")

move_ship <- function(directions) {
  xpos = 0
  ypos = 0
  angle = 0
  dx <- c("0" = 1, "90" = 0, "180" = -1, "270" = 0)
  dy <- c("0" = 0, "90" = 1, "180" = 0, "270" = -1)
  for (i in seq_along(directions)) {
    amount <- as.numeric(gsub("[A-Z]", "", directions[i]))
    command <- gsub("[0-9]+", "", directions[i])
    if (command == "N") {
      ypos <- ypos + amount
      next
    }
    if (command == "S") {
      ypos <- ypos - amount
      next
    }
    if (command == "E") {
      xpos <- xpos + amount
      next
    }
    if (command == "W") {
      xpos <- xpos - amount
      next
    }
    if (command == "R") {
      angle <- (angle + amount) %% 360
      next
    }
    if (command == "L") {
      angle <- (angle - amount) %% 360
      next
    }
    if (command == "F") {
      xpos <- xpos + dx[[as.character(angle)]] * amount
      ypos <- ypos + dy[[as.character(angle)]] * amount
      next
    }
  }
  c(xpos, ypos)
}
